src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+        self.kwargs = kwargs
+
+        base_layer_mod = self.get_base_layer()
+        if isinstance(base_layer_mod, nn.Linear):
+            self.in_features, self.out_features = base_layer_mod.in_features, base_layer_mod.out_features
+        else:
+            raise ValueError(f"Unsupported layer type {type(base_layer_mod)}")
+
+    @staticmethod
+    def _compute_delta_from_params(
+        A: torch.Tensor, B: torch.Tensor, alpha: torch.Tensor, r: int, w_norm: torch.Tensor
+    ) -> torch.Tensor:
+        """Compute delta = B @ diag(alpha/r / (||A_i||*||B^j||)) @ A, scaled by provided w_norm (per-input channel)"""
+        comp_dtype = torch.float32 if A.dtype in (torch.float16, torch.bfloat16) else A.dtype
+        A_ = A.to(comp_dtype)
+        B_ = B.to(comp_dtype)
+        alpha_ = alpha.to(comp_dtype)
+        w_norm_ = w_norm.to(comp_dtype)
+        An = torch.clamp(A_.norm(dim=1), min=1e-12)
+        Bn = torch.clamp(B_.norm(dim=0), min=1e-12)
+        diag = torch.diag_embed((alpha_ / r) / (An * Bn))
+        delta = B_ @ diag @ A_
+        delta = delta * w_norm_.unsqueeze(0)
+        return delta.to(A.dtype, copy=False)
+
+    def get_delta_weight(self, adapter: str) -> torch.Tensor:
+        if adapter not in self.delora_A or adapter not in self.delora_B:
+            raise ValueError(f"Adapter {adapter} not found.")
+        base_w = self.get_base_layer().weight
+        out_dtype = base_w.dtype
+        device = base_w.device
+        comp_dtype = torch.float32 if out_dtype in (torch.float16, torch.bfloat16) else out_dtype
+
+        A = self.delora_A[adapter].to(dtype=comp_dtype, device=device)
+        B = self.delora_B[adapter].to(dtype=comp_dtype, device=device)
+        alpha = self.delora_alpha[adapter].to(dtype=comp_dtype, device=device)
+        An = torch.clamp(A.norm(dim=1), min=1e-12)
+        Bn = torch.clamp(B.norm(dim=0), min=1e-12)
+        diag = torch.diag_embed((alpha / self.r[adapter]) / (An * Bn))  # (r,r)
+        delta = B @ diag @ A  # (out,in)
+
+        if (not self.use_residual_init) and (adapter in self.delora_initial_A and adapter in self.delora_initial_B):
+            A0 = self.delora_initial_A[adapter].to(dtype=comp_dtype, device=device)
+            B0 = self.delora_initial_B[adapter].to(dtype=comp_dtype, device=device)
+            alpha0 = self.delora_initial_alpha[adapter].to(dtype=comp_dtype, device=device).squeeze(0)
+            A0n = torch.clamp(A0.norm(dim=1), min=1e-12)
+            B0n = torch.clamp(B0.norm(dim=0), min=1e-12)
+            diag0 = torch.diag_embed((alpha0 / self.r[adapter]) / (A0n * B0n))
+            delta = delta - (B0 @ diag0 @ A0)
+        # scale by a fixed per-adapter reference base weight norm when available (avoids caching for exact unmerge)
+        if adapter in self.delora_w_norm:
+            w_norm = self.delora_w_norm[adapter].to(dtype=comp_dtype, device=device)
+        else:
+            # fallback: compute current base weight norm on-the-fly
```

Comment:
Under what conditions could `self.delora_w_norm` not contain the adapter?

Reply:
Indeed this was mistakenly still in the code. I now removed it